File: src/sdk/cloud-py/ratel_ai_cloud/canonical.py
"""Frozen v1 canonicalization — the Python mirror of the reference
implementation in `protocol/v1/conformance/verify.mjs`.

The ETag content projection is frozen at v1 (see `protocol/v1/README.md`):
exactly the seven wire fields in a fixed key order, `metadata` keys sorted by
UTF-8 byte order, arrays in authored order, raw UTF-8, compact JSON, set sorted
by `id`. Runtime uses this for nothing but tests and the mock source — a loader
trusts the server's ETag; recomputation is a test-only integrity check.
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Mapping, Sequence
from typing import Any, NamedTuple

__all__ = ["SKILL_FIELDS", "Etag", "canonical_set", "canonical_skill", "etag_of", "resolve"]

# The frozen v1 content projection: exactly these fields, in this order.
SKILL_FIELDS = ("id", "name", "description", "tags", "tools", "metadata", "body")


class Etag(NamedTuple):
    hex: str
    etag: str  # the strong entity-tag: `"<hex>"`


def _json(value: Any) -> str:
    # Compact separators + ensure_ascii=False match JSON.stringify: minimal
    # escaping, raw UTF-8, no insignificant whitespace.
    return json.dumps(value, separators=(",", ":"), ensure_ascii=False)


def _utf8(key: str) -> bytes:
    return key.encode("utf-8")


def canonical_skill(skill: Mapping[str, Any]) -> str:
    """Canonical JSON for one skill, projected to exactly the seven wire fields."""
    metadata: Mapping[str, Any] = skill.get("metadata") or {}
    meta_keys = sorted(metadata, key=_utf8)
    meta = "{" + ",".join(_json(k) + ":" + _json(metadata[k]) for k in meta_keys) + "}"
    return (
        "{"
        + '"id":' + _json(skill["id"]) + ","
        + '"name":' + _json(skill["name"]) + ","
        + '"description":' + _json(skill["description"]) + ","
        + '"tags":' + _json(skill["tags"]) + ","
        + '"tools":' + _json(skill["tools"]) + ","
        + '"metadata":' + meta + ","
        + '"body":' + _json(skill["body"])
        + "}"
    )


def _sort_by_id(skills: Sequence[Mapping[str, Any]]) -> list[Mapping[str, Any]]:
    return sorted(skills, key=lambda s: _utf8(s["id"]))


def canonical_set(skills: Sequence[Mapping[str, Any]]) -> str:
    """Canonical bytes for a resolved set: sorted by id, compact JSON array."""
    return "[" + ",".join(canonical_skill(s) for s in _sort_by_id(skills)) + "]"


def etag_of(skills: Sequence[Mapping[str, Any]]) -> Etag:
    hex_digest = hashlib.sha256(canonical_set(skills).encode("utf-8")).hexdigest()
    return Etag(hex=hex_digest, etag=f'"{hex_digest}"')


def resolve(catalog: Mapping[str, Any], scope: str | None) -> list[Mapping[str, Any]]:
    """Resolve the published set for a scope: absent scope => the global layer;
    a subject => its layer overlaid on global, subject winning on `name`
    collision; an unknown subject => the global layer (empty overlay)."""
    global_layer: Sequence[Mapping[str, Any]] = catalog.get("global") or []
    if scope is None:
        return _sort_by_id(global_layer)
    subject_layer: Sequence[Mapping[str, Any]] = (catalog.get("subjects") or {}).get(scope) or []
    by_name: dict[str, Mapping[str, Any]] = {}
    for skill in global_layer:
        by_name[skill["name"]] = skill
    for skill in subject_layer:
        by_name[skill["name"]] = skill
    return _sort_by_id(list(by_name.values()))

File: src/sdk/cloud/src/canonical.ts
/**
 * The frozen protocol/v1 content projection and ETag algorithm
 * (`protocol/v1/README.md`, pinned by `protocol/v1/conformance/vectors.json`).
 * Reimplemented here — the reference verifier is not importable — and pinned
 * against the committed vectors in `canonical.test.ts`. Changing any byte of
 * this serialization is a protocol v2.
 */

import { createHash } from "node:crypto";

/** The wire shape of one skill in a v1 catalog pull (`catalog-skill.schema.json`). */
export interface CatalogSkillWire {
  id: string;
  name: string;
  description: string;
  tags: string[];
  tools: string[];
  metadata: Record<string, string[]>;
  body: string;
}

/**
 * Anything projectable to the wire shape: the SDK `Skill` (optional fields) and
 * any raw source-side object carrying extra fields the projection drops.
 */
export interface SkillLike {
  id: string;
  name: string;
  description: string;
  tags?: string[];
  tools?: string[];
  metadata?: Record<string, string[]>;
  body?: string;
}

/** Source-side layers a scope resolves against: one global set plus per-subject overlays. */
export interface SourceLayers {
  global: CatalogSkillWire[];
  subjects?: Record<string, CatalogSkillWire[]>;
}

// UTF-8 bytewise comparison, for the set sort-by-id and the metadata key sort.
function byteCompare(a: string, b: string): number {
  return Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));
}

/** Project a skill to exactly the 7 wire fields, defaulting the SDK's optional ones. */
export function projectSkill(skill: SkillLike): CatalogSkillWire {
  return {
    id: skill.id,
    name: skill.name,
    description: skill.description,
    tags: skill.tags ?? [],
    tools: skill.tools ?? [],
    metadata: skill.metadata ?? {},
    body: skill.body ?? "",
  };
}

/**
 * Canonical JSON for one projected skill: the 7 keys in fixed order, metadata
 * keys byte-sorted, arrays in authored order, minimal escaping, raw UTF-8, no
 * whitespace. `JSON.stringify` on strings/arrays already emits minimal escapes
 * and raw UTF-8; key order is driven by hand so object iteration order can
 * never influence the bytes.
 */
export function canonicalSkill(skill: SkillLike): string {
  const p = projectSkill(skill);
  const s = JSON.stringify;
  const metaKeys = Object.keys(p.metadata).sort(byteCompare);
  const meta = `{${metaKeys.map((k) => `${s(k)}:${s(p.metadata[k])}`).join(",")}}`;
  return (
    `{"id":${s(p.id)},"name":${s(p.name)},"description":${s(p.description)},` +
    `"tags":${s(p.tags)},"tools":${s(p.tools)},"metadata":${meta},"body":${s(p.body)}}`
  );
}

/** Canonical bytes for a resolved set: skills sorted by id, compact JSON array. */
export function canonicalSet(skills: SkillLike[]): string {
  const sorted = [...skills].sort((a, b) => byteCompare(a.id, b.id));
  return `[${sorted.map(canonicalSkill).join(",")}]`;
}

/** ETag over a resolved set: lowercase hex sha256 plus the quoted strong tag. */
export function etagOf(skills: SkillLike[]): { hex: string; etag: string } {
  const hex = createHash("sha256").update(canonicalSet(skills), "utf8").digest("hex");
  return { hex, etag: `"${hex}"` };
}

/**
 * Resolve the published set for a scope: absent scope or unknown subject means
 * the global layer only; a known subject overlays its layer on the global one,
 * the subject winning on `name` collision (the merge key is name, not id).
 * The merged set is re-sorted by id.
 */
export function resolveScope(layers: SourceLayers, scope?: string | null): CatalogSkillWire[] {
  const global = layers.global ?? [];
  const sortById = (skills: CatalogSkillWire[]) =>
    [...skills].sort((a, b) => byteCompare(a.id, b.id));
  if (scope == null) return sortById(global);
  const subject = layers.subjects?.[scope] ?? [];
  const byName = new Map<string, CatalogSkillWire>();
  for (const sk of global) byName.set(sk.name, sk);
  for (const sk of subject) byName.set(sk.name, sk);
  return sortById([...byName.values()]);
}

/** True when two skills agree on exactly the 7 wire fields — the sync idempotence gate. */
export function skillsEqual(a: SkillLike, b: SkillLike): boolean {
  return canonicalSkill(a) === canonicalSkill(b);
}
